File: arduinobot_firmware/arduinobot_firmware/simple_serial_receiver.py
#!/usr/bin/env python3
import rclpy
import serial
from rclpy.node import Node
from std_msgs.msg import String

class SimpleSerialReceiver(Node):
    def __init__(self):
        super().__init__("simple_serial_receiver")
        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baud_rate", 115200)
        self.port_ = self.get_parameter("port").value
        self.baud_rate_ = self.get_parameter("baud_rate").value
        
        
        self.frequency_ = 0.01
        self.arduinobot_ = serial.Serial(port=self.port_, baudrate=self.baud_rate_, timeout=0.1)
        self.pub_ = self.create_publisher(String, "serial_receiver", 10)
        self.timer_ = self.create_timer(self.frequency_, self.timerCallback)

    def timerCallback(self):
        if rclpy.ok() and self.arduinobot_.is_open:
            data = self.arduinobot_.readline()
            try:
                data.decode("utf-8")
            except:
                return
            
            msg = String()
            msg.data = str(data)
        
        self.pub_.publish(msg)

def main():
    rclpy.init()
    node = SimpleSerialReceiver()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: arduinobot_firmware/src/simple_serial_receiver.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "libserial/SerialPort.h"

using namespace std::chrono_literals;

class SimpleSerialReceiver : public rclcpp::Node
{
public:
    SimpleSerialReceiver() : Node("simple_serial_receiver")
    {
        declare_parameter<std::string>("port", "/dev/ttyUSB0");
        std::string port_ = get_parameter("port").as_string();

        publisher_ = this->create_publisher<std_msgs::msg::String>("serial_receiver", 10);
        timer_ = this->create_wall_timer(0.01s, std::bind(&SimpleSerialReceiver::timer_callback, this));
        arudino_.Open(port_);
        arudino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);
    }
    ~SimpleSerialReceiver()
    {
        arudino_.Close();
    }
private:
    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        if (rclcpp::ok() && arudino_.IsDataAvailable())
        {
            arudino_.ReadLine(message.data);
        }
        publisher_->publish(message);
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    LibSerial::SerialPort arudino_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SimpleSerialReceiver>());
  rclcpp::shutdown();
  return 0;
}